### PR TITLE
swat port for t/* stuff

### PR DESCRIPTION
**Pull request by melezhik**
Hi @Manwar! 

[swat](https://github.com/melezhik/swat) is a test automation framework. 

I found it quite interesting to port some of t/* stuff to swat DSL. On my laptop output looks like this:

```
vagrant@Debian-jessie-amd64-netboot:~/projects/Map-Tube-London$ carton exec swat swat-test/
/home/vagrant/.swat/.cache/8803/prove/route/00.GET.t .........
ok 1 - response is already set
# response saved to /home/vagrant/.swat/.cache/8803/prove/S3gLorL1z0
ok 2 - GET /route returns Route 1 Baker Street => Farringdon : Baker Street (Circle, Hammersmith & City, Metropolitan), Great Portland Street (Circle, Hammersmith & City, Metropolitan), Euston Square (Circle, Hammersmith & City, Metropolitan), King's Cross St. Pancras (Circle, Hammersmith & City, Metropolitan), Farringdon (Circle, Hammersmith & City, Metropolitan)
ok 3 - GET /route returns Route 2 Bank => Monument : Bank (Tunnel), Monument (Tunnel)
ok 4 - GET /route returns Route 3 Euston => King's Cross St. Pancras : Euston (Northern, Street, Victoria), King's Cross St. Pancras (Northern, Street, Victoria)
1..4
ok
/home/vagrant/.swat/.cache/8803/prove/test-case-01/00.GET.t ..
ok 1 - response is already set
# response saved to /home/vagrant/.swat/.cache/8803/prove/2izmIDf9GG
ok 2 - GET /test-case-01 returns | Route 1 found: Wembley Central (Overground, Bakerloo), Stonebridge Park (Overground, Bakerloo), Harlesden (Overground, Bakerloo), Willesdon Junction (Overground, Bakerloo), Shepherd's Bush (Central, Overground), Holland Park (Central), Notting Hill Gate (Central, Circle, District), Queensway (Central), Lancaster Gate (Central), Marble Arch (Central), Bond Street (Central, Jubilee)
ok 3 - GET /test-case-01 returns | Route 1 expected: Wembley Central (Overground, Bakerloo), Stonebridge Park (Overground, Bakerloo), Harlesden (Overground, Bakerloo), Willesdon Junction (Overground, Bakerloo), Shepherd's Bush (Central, Overground), Holland Park (Central), Notting Hill Gate (Central, Circle, District), Queensway (Central), Lancaster Gate (Central), Marble Arch (Central), Bond Street (Central, Jubilee)
ok 4 - GET /test-case-01 returns | Route 2 found: Bond Street (Central, Jubilee), Oxford Circus (Central, Victoria, Bakerloo), Warren Street (Northern, Victoria), Euston (Northern, Overground, Victoria, Street)
ok 5 - GET /test-case-01 returns | Route 2 expected: Bond Street (Central, Jubilee), Oxford Circus (Central, Victoria, Bakerloo), Warren Street (Northern, Victoria), Euston (Northern, Overground, Victoria, Street)
ok 6 - GET /test-case-01 returns | Route 3 found: White City (Central), Shepherd's Bush (Central, Overground), Kensington (Olympia) (District, Overground), Earl's Court (District, Piccadilly), Gloucester Road (Circle, District, Piccadilly), South Kensington (Circle, District, Piccadilly), Sloane Square (Circle, District), Victoria (Circle, District, Victoria)
ok 7 - GET /test-case-01 returns | Route 3 expected: White City (Central), Shepherd's Bush (Central, Overground), Kensington (Olympia) (District, Overground), Earl's Court (District, Piccadilly), Gloucester Road (Circle, District, Piccadilly), South Kensington (Circle, District, Piccadilly), Sloane Square (Circle, District), Victoria (Circle, District, Victoria)
ok 8 - GET /test-case-01 returns | Route 4 found: Temple (Circle, District), Embankment (Circle, District, Northern, Bakerloo), Waterloo (Northern, Bakerloo, Jubilee, Waterloo & City), Bank (Central, DLR, Northern, Waterloo & City, Tunnel), Moorgate (Circle, Northern, Hammersmith & City, Metropolitan), Barbican (Circle, Hammersmith & City, Metropolitan), Farringdon (Circle, Hammersmith & City, Metropolitan)
ok 9 - GET /test-case-01 returns | Route 4 expected: Temple (Circle, District), Embankment (Circle, District, Northern, Bakerloo), Waterloo (Northern, Bakerloo, Jubilee, Waterloo & City), Bank (Central, DLR, Northern, Waterloo & City, Tunnel), Moorgate (Circle, Northern, Hammersmith & City, Metropolitan), Barbican (Circle, Hammersmith & City, Metropolitan), Farringdon (Circle, Hammersmith & City, Metropolitan)
ok 10 - GET /test-case-01 returns | Route 5 found: Turnham Green (District, Piccadilly), Hammersmith (Circle, District, Hammersmith & City, Piccadilly), Barons Court (District, Piccadilly), Earl's Court (District, Piccadilly), Gloucester Road (Circle, District, Piccadilly), South Kensington (Circle, District, Piccadilly), Knightsbridge (Piccadilly), Hyde Park Corner (Piccadilly), Green Park (Victoria, Jubilee, Piccadilly), Westminster (Circle, District, Jubilee), Waterloo (Northern, Bakerloo, Jubilee, Waterloo & City), Bank (Central, DLR, Northern, Waterloo & City, Tunnel), Shadwell (DLR, Overground), Whitechapel (District, Hammersmith & City, Overground)
ok 11 - GET /test-case-01 returns | Route 5 expected: Turnham Green (District, Piccadilly), Hammersmith (Circle, District, Hammersmith & City, Piccadilly), Barons Court (District, Piccadilly), Earl's Court (District, Piccadilly), Gloucester Road (Circle, District, Piccadilly), South Kensington (Circle, District, Piccadilly), Knightsbridge (Piccadilly), Hyde Park Corner (Piccadilly), Green Park (Victoria, Jubilee, Piccadilly), Westminster (Circle, District, Jubilee), Waterloo (Northern, Bakerloo, Jubilee, Waterloo & City), Bank (Central, DLR, Northern, Waterloo & City, Tunnel), Shadwell (DLR, Overground), Whitechapel (District, Hammersmith & City, Overground)
ok 12 - GET /test-case-01 returns | Route 6 found: Goldhawk Road (Circle, Hammersmith & City), Shepherd's Bush Market (Circle, Hammersmith & City), Wood Lane (Circle, Hammersmith & City), White City (Central), Shepherd's Bush (Central, Overground), Willesdon Junction (Overground, Bakerloo), Harlesden (Overground, Bakerloo), Stonebridge Park (Overground, Bakerloo), Wembley Central (Overground, Bakerloo)
ok 13 - GET /test-case-01 returns | Route 6 expected: Goldhawk Road (Circle, Hammersmith & City), Shepherd's Bush Market (Circle, Hammersmith & City), Wood Lane (Circle, Hammersmith & City), White City (Central), Shepherd's Bush (Central, Overground), Willesdon Junction (Overground, Bakerloo), Harlesden (Overground, Bakerloo), Stonebridge Park (Overground, Bakerloo), Wembley Central (Overground, Bakerloo)
ok 14 - GET /test-case-01 returns | Route 7 found: Wembley Central (Overground, Bakerloo), Stonebridge Park (Overground, Bakerloo), Harlesden (Overground, Bakerloo), Willesdon Junction (Overground, Bakerloo), Shepherd's Bush (Central, Overground), Holland Park (Central), Notting Hill Gate (Central, Circle, District), Bayswater (Circle, District), Paddington (District, Circle, Hammersmith & City, Bakerloo), Edgware Road (Circle, District, Hammersmith & City, Bakerloo), Marylebone (Bakerloo)
ok 15 - GET /test-case-01 returns | Route 7 expected: Wembley Central (Overground, Bakerloo), Stonebridge Park (Overground, Bakerloo), Harlesden (Overground, Bakerloo), Willesdon Junction (Overground, Bakerloo), Shepherd's Bush (Central, Overground), Holland Park (Central), Notting Hill Gate (Central, Circle, District), Bayswater (Circle, District), Paddington (District, Circle, Hammersmith & City, Bakerloo), Edgware Road (Circle, District, Hammersmith & City, Bakerloo), Marylebone (Bakerloo)
ok 16 - GET /test-case-01 returns | Route 8 found: Baker Street (Circle, Hammersmith & City, Bakerloo, Metropolitan, Jubilee), Finchley Road (Metropolitan, Jubilee), Wembley Park (Metropolitan, Jubilee), Neasden (Jubilee)
ok 17 - GET /test-case-01 returns | Route 8 expected: Baker Street (Circle, Hammersmith & City, Bakerloo, Metropolitan, Jubilee), Finchley Road (Metropolitan, Jubilee), Wembley Park (Metropolitan, Jubilee), Neasden (Jubilee)
ok 18 - GET /test-case-01 returns | Route 9 found: Baker Street (Circle, Hammersmith & City, Bakerloo, Metropolitan, Jubilee), Finchley Road (Metropolitan, Jubilee), Wembley Park (Metropolitan, Jubilee), Preston Road (Metropolitan)
ok 19 - GET /test-case-01 returns | Route 9 expected: Baker Street (Circle, Hammersmith & City, Bakerloo, Metropolitan, Jubilee), Finchley Road (Metropolitan, Jubilee), Wembley Park (Metropolitan, Jubilee), Preston Road (Metropolitan)
ok 20 - GET /test-case-01 returns | Route 10 found: Oval (Northern), Kennington (Northern), Waterloo (Northern, Bakerloo, Jubilee, Waterloo & City), Westminster (Circle, District, Jubilee), Green Park (Victoria, Jubilee, Piccadilly), Oxford Circus (Central, Victoria, Bakerloo), Warren Street (Northern, Victoria), Euston (Northern, Overground, Victoria, Street)
ok 21 - GET /test-case-01 returns | Route 10 expected: Oval (Northern), Kennington (Northern), Waterloo (Northern, Bakerloo, Jubilee, Waterloo & City), Westminster (Circle, District, Jubilee), Green Park (Victoria, Jubilee, Piccadilly), Oxford Circus (Central, Victoria, Bakerloo), Warren Street (Northern, Victoria), Euston (Northern, Overground, Victoria, Street)
ok 22 - GET /test-case-01 returns | Route 11 found: South Ealing (Piccadilly), Acton Town (District, Piccadilly), Ealing Common (District, Piccadilly), North Ealing (Piccadilly), Park Royal (Piccadilly), Alperton (Piccadilly)
ok 23 - GET /test-case-01 returns | Route 11 expected: South Ealing (Piccadilly), Acton Town (District, Piccadilly), Ealing Common (District, Piccadilly), North Ealing (Piccadilly), Park Royal (Piccadilly), Alperton (Piccadilly)
ok 24 - GET /test-case-01 returns | Route 12 found: Bank (Central, DLR, Northern, Waterloo & City, Tunnel), Waterloo (Northern, Bakerloo, Jubilee, Waterloo & City), Westminster (Circle, District, Jubilee)
ok 25 - GET /test-case-01 returns | Route 12 expected: Bank (Central, DLR, Northern, Waterloo & City, Tunnel), Waterloo (Northern, Bakerloo, Jubilee, Waterloo & City), Westminster (Circle, District, Jubilee)
ok 26 - GET /test-case-01 returns | Route 13 found: Hoxton (Overground), Haggerston (Overground), Dalston Junction (Overground), Dalston Kingsland (Overground), Canonbury (Overground), Highbury & Islington (Overground, Victoria), Caledonian Road & Barnsbury (Overground), Camden Road (Overground), Kentish Town West (Overground), Gospel Oak (Overground)
ok 27 - GET /test-case-01 returns | Route 13 expected: Hoxton (Overground), Haggerston (Overground), Dalston Junction (Overground), Dalston Kingsland (Overground), Canonbury (Overground), Highbury & Islington (Overground, Victoria), Caledonian Road & Barnsbury (Overground), Camden Road (Overground), Kentish Town West (Overground), Gospel Oak (Overground)
ok 28 - GET /test-case-01 returns | Route 14 found: Baker Street (Circle, Hammersmith & City, Bakerloo, Metropolitan, Jubilee), Finchley Road (Metropolitan, Jubilee), Harrow-on-the-Hill (Metropolitan), North Harrow (Metropolitan)
ok 29 - GET /test-case-01 returns | Route 14 expected: Baker Street (Circle, Hammersmith & City, Bakerloo, Metropolitan, Jubilee), Finchley Road (Metropolitan, Jubilee), Harrow-on-the-Hill (Metropolitan), North Harrow (Metropolitan)
ok 30 - GET /test-case-01 returns | Route 15 found: Baker Street (Circle, Hammersmith & City, Bakerloo, Metropolitan, Jubilee), Finchley Road (Metropolitan, Jubilee), Harrow-on-the-Hill (Metropolitan), Moor Park (Metropolitan), Croxley (Metropolitan)
ok 31 - GET /test-case-01 returns | Route 15 expected: Baker Street (Circle, Hammersmith & City, Bakerloo, Metropolitan, Jubilee), Finchley Road (Metropolitan, Jubilee), Harrow-on-the-Hill (Metropolitan), Moor Park (Metropolitan), Croxley (Metropolitan)
ok 32 - GET /test-case-01 returns | Route 16 found: Sloane Square (Circle, District), Victoria (Circle, District, Victoria), St. James's Park (Circle, District), Westminster (Circle, District, Jubilee)
ok 33 - GET /test-case-01 returns | Route 16 expected: Sloane Square (Circle, District), Victoria (Circle, District, Victoria), St. James's Park (Circle, District), Westminster (Circle, District, Jubilee)
ok 34 - GET /test-case-01 returns | Route 17 found: Westferry (DLR), Limehouse (DLR), Shadwell (DLR, Overground), Bank (Central, DLR, Northern, Waterloo & City, Tunnel), Monument (Circle, District, Tunnel), Cannon Street (Circle, District)
ok 35 - GET /test-case-01 returns | Route 17 expected: Westferry (DLR), Limehouse (DLR), Shadwell (DLR, Overground), Bank (Central, DLR, Northern, Waterloo & City, Tunnel), Monument (Circle, District, Tunnel), Cannon Street (Circle, District)
ok 36 - GET /test-case-01 returns | Route 18 found: Westferry (DLR), Limehouse (DLR), Shadwell (DLR, Overground), Bank (Central, DLR, Northern, Waterloo & City, Tunnel), Monument (Circle, District, Tunnel), Cannon Street (Circle, District)
ok 37 - GET /test-case-01 returns | Route 18 expected: Westferry (DLR), Limehouse (DLR), Shadwell (DLR, Overground), Bank (Central, DLR, Northern, Waterloo & City, Tunnel), Monument (Circle, District, Tunnel), Cannon Street (Circle, District)
ok 38 - GET /test-case-01 returns | Route 19 found: Westferry (DLR), Limehouse (DLR), Shadwell (DLR, Overground), Bank (Central, DLR, Northern, Waterloo & City, Tunnel), Monument (Circle, District, Tunnel), Cannon Street (Circle, District)
ok 39 - GET /test-case-01 returns | Route 19 expected: Westferry (DLR), Limehouse (DLR), Shadwell (DLR, Overground), Bank (Central, DLR, Northern, Waterloo & City, Tunnel), Monument (Circle, District, Tunnel), Cannon Street (Circle, District)
ok 40 - GET /test-case-01 returns | Route 20 found: Westferry (DLR), Limehouse (DLR), Shadwell (DLR, Overground), Bank (Central, DLR, Northern, Waterloo & City, Tunnel), Monument (Circle, District, Tunnel), Cannon Street (Circle, District)
ok 41 - GET /test-case-01 returns | Route 20 expected: Westferry (DLR), Limehouse (DLR), Shadwell (DLR, Overground), Bank (Central, DLR, Northern, Waterloo & City, Tunnel), Monument (Circle, District, Tunnel), Cannon Street (Circle, District)
ok 42 - GET /test-case-01 returns | Route 21 found: Westferry (DLR), Limehouse (DLR), Shadwell (DLR, Overground), Bank (Central, DLR, Northern, Waterloo & City, Tunnel), Monument (Circle, District, Tunnel), Cannon Street (Circle, District)
ok 43 - GET /test-case-01 returns | Route 21 expected: Westferry (DLR), Limehouse (DLR), Shadwell (DLR, Overground), Bank (Central, DLR, Northern, Waterloo & City, Tunnel), Monument (Circle, District, Tunnel), Cannon Street (Circle, District)
ok 44 - GET /test-case-01 returns | Route 22 found: Westferry (DLR), Limehouse (DLR), Shadwell (DLR, Overground), Bank (Central, DLR, Northern, Waterloo & City, Tunnel), Monument (Circle, District, Tunnel), Cannon Street (Circle, District)
ok 45 - GET /test-case-01 returns | Route 22 expected: Westferry (DLR), Limehouse (DLR), Shadwell (DLR, Overground), Bank (Central, DLR, Northern, Waterloo & City, Tunnel), Monument (Circle, District, Tunnel), Cannon Street (Circle, District)
ok 46 - GET /test-case-01 returns | Route 23 found: Tower Gateway (DLR, Street), Tower Hill (Circle, District, Street), Aldgate (Circle, Metropolitan)
ok 47 - GET /test-case-01 returns | Route 23 expected: Tower Gateway (DLR, Street), Tower Hill (Circle, District, Street), Aldgate (Circle, Metropolitan)
ok 48 - GET /test-case-01 returns | Route 24 found: Liverpool Street (Central, Circle, Hammersmith & City, Metropolitan), Bank (Central, DLR, Northern, Waterloo & City, Tunnel), Monument (Circle, District, Tunnel)
ok 49 - GET /test-case-01 returns | Route 24 expected: Liverpool Street (Central, Circle, Hammersmith & City, Metropolitan), Bank (Central, DLR, Northern, Waterloo & City, Tunnel), Monument (Circle, District, Tunnel)
1..49
ok
/home/vagrant/.swat/.cache/8803/prove/test-case-00/00.GET.t ..
ok 1 - response is already set
# response saved to /home/vagrant/.swat/.cache/8803/prove/DOCTZmp8kg
ok 2 - GET /test-case-00 returns | get_shortest_route() : Map::Tube::get_shortest_route(): ERROR: Either FROM/TO node is undefined
ok 3 - GET /test-case-00 returns | get_shortest_route('Bond Street') : Map::Tube::get_shortest_route(): ERROR: Either FROM/TO node is undefined
ok 4 - GET /test-case-00 returns | get_shortest_route('XYZ','Bond Street') : Map::Tube::get_node_by_name(): ERROR: Invalid station name
ok 5 - GET /test-case-00 returns | get_shortest_route('Bond Street', 'XYZ') : Map::Tube::get_node_by_name(): ERROR: Invalid station name
1..5
ok
All tests successful.
Files=3, Tests=58,  2 wallclock secs ( 0.03 usr  0.01 sys +  1.18 cusr  0.02 csys =  1.24 CPU)
Result: PASS
```

If that sounds interesting for you we could talk about further co-working. 

The main idea is to instead of using `is_deep` from Test::More I would suggest a swat DSL to verify various text output.

 At least because it is much easier to implement and less fragile , as long as one holds a dedicated text format emitted to  swat , which is not a big deal. Also it might be consider as black box testing as this approach less copulated with inner object structures.

### CODE REVIEW

**Comment by melezhik** [comment]
and, yes to run a tests just say: `carton exec swat swat-test/`

**Comment by Manwar** [comment]
Hi @melezhik,

Thanks for your contribution, much appreciated.

I must confess that I have never heard of 'swat' before. This is my first introduction to the testing tool. I like the philosophy behind it. However I find it difficult to follow. It could be just me. I am being honest with you.

I like your offer of working together. I am always ready for it. I have a suggestion for you, though. I like to keep test code and source code separate, to keep it light weight. If you have noticed the package Map::Tube::London is very thin wrapper around the package Map::Tube. So adding 'swat' is going to be quite heavy, in my humble opinion. Don't get me wrong, I am not against it. 

Instead I suggest, we (both) can add 'swat' to the package Test::Map::Tube, which currently expose one method ok_map(). It validates the map data. I would suggest add new method ok_map_features() to this package that would use 'swat' to do all kind of unit testing. 

This way other users of Map::Tube can also benefit without any overhead. What do you think?

Please share your view.

Many Thanks.

Best Regards,
Mohammad S Anwar

**Comment by melezhik** [comment]
Hi Mohammad! Thanks alot for your feedback. I understand your concerns
about dependencies overheads. Indeed swat is standalone tool and not to be
used as Perl module, it acts as Perl prove util though and produces TAP
output as well. There are probably two variants we choose between.

First try to ship swat test together with map-tube*modules and run swat
tests after unit tests , but this requires modification of existing
makefile.pl or build.pl and hacking into distribution installation process.

Second way is to keep swat tests separately from source code and run them
separately from unit tests, this means only developers not end users could
run such a tests and they won't be run during cpan installation process.

Both ways have pros and cons, whatever way we choose I want to emphasize an
idea behind swat and map-tube.

I consider map tube system as black box one, the way I verify it works
correctly is a verifying of text output provided by map tube methods, like
shortest_route one - It is quite  convenient to map subway routes into
a chains of subway stations names represented by a regular line of text:
"foo, bar, baz, etc." and then "grep" inside this line , and this is where
swat fits the best, as it is text oriented test tool ( of course swat
could do more ... )

So the thing I guess we really need to focus in is real life use cases
could be tested in terms of swat, and this is where you definitely knows
much more than me.

PS btw , I am just curios - what is practical meaning of all theses
map-tube* modules ? Who use it ?

**Comment by melezhik** [comment]
Hi Mohammad ! Any news here? Just want to hear what do you think. Thanks,

**Comment by Manwar** [comment]
Hi Alexey,

Sorry for late response, got distracted in other projects. I am not sure if I understood everything you said. I would like you to take a lead role in this new project of incorporating swat with 'Map::Tube'. The only thing I am concerned that it should be optional for end user. I am not sure how this can be achieved but I will leave that to your expertise. I am happy to answer any questions you have in the process. In the meantime, I will try to get myself familiarize with 'swat' tool. 

Do I make any sense?

Many Thanks.

Best Regards,
Mohammad S Anwar

**Comment by melezhik** [comment]
Hi Mohammad ! Thanks for your answer. After some mediation I decided to postpone further activity on this , I want to create a brand new test framework - [outhentic](https://github.com/melezhik/outhentic)  which is inspired by swat but pure generic ( while swat is web testing one ) and fits much better for map-tube-* testing, when I have done I will let you know. Thanks

**Comment by Manwar** [comment]
Hi Alexey,

That's great news !!!
Is there anything I can be of any help, please let me know.

Best Regards,
Mohammad S Anwar

**Comment by melezhik** [comment]
Hi!
> Is there anything I can be of any help, please let me know.

Once I asked, "I am just curios - what is practical meaning of all theses map-tube* modules ? Who use it ?" 

Thanks (

**Comment by Manwar** [comment]
Hi Alexey,

Nearly 5 years, 2010-10-25, to be precise, I uploaded my first draft of Map::Tube to CPAN. It started as a fun project and to put something in public domain. It has been through many fold of design changes since then, thanks to many people. 

It was meant to deal with London tube map, which is the life line of many of us, here in London. I had no previous experience of AI, this was a big task for me. Currently there are four of us, who is working on this project, namely Michal Špaček (SKIM), Gisbert W. Selke (GWS) and Slaven Rezić (SREZIC).

There is no real purpose of this project at the moment. However I am planning to create REST API to expose the functionalities and host it on my website www.manwar.org. And even create a page to let the online user play with it. All for fun.

Looking at the GitHub/MetaCPAN, it appears few do like the project, which is encouraging.

Best Regards,
Mohammad S Anwar

**Comment by melezhik** [comment]
Hi Mohammad! I have just released https://github.com/melezhik/outthentic to CPAN. I guess it might be useful in map tube tests. I think we could close this one. I am going to rewrite my initial merge request using new framework and then I will MR to you again.

**Comment by Manwar** [comment]
Hi Alexey, 

I will get back to you once I look through the details.
All the best.

Best Regards,
Mohammad S Anwar

**Comment by melezhik** [comment]
Sure , Mohammad.

**Comment by melezhik** [comment]
Hi Mohammad!
I have created a simple test suite for map-tube family modules -  https://github.com/melezhik/map-tube-testers .  You may play with it. It based on [outthentic](https://github.com/melezhik/outthentic) - a new test framework with the DSL very like to swat one. Eventually if it sounds interesting a new tests / maps may popup. Please take a look and let me know what do you think. Thanks.

**Comment by Manwar** [comment]
Hi Alexey,

Thanks for the update.
I find no documentation whatsoever. It would really encourage, if you had explained the design of your tool. Also if you could throw some lights on how it is better than traditional approach. Although you have plenty of examples, I find it hard to follow.

Best Regards,
Mohammad S Anwar

**Comment by melezhik** [comment]
Hi Mohammad. Outthentic documentaion is here - https://github.com/melezhik/outthentic ,  the link I sent you ( https://github.com/melezhik/map-tube-testers ) is an example how o

**Comment by melezhik** [comment]
is an example how outtentic might be used in map tube project. To understand the difference against a common test frame works , it is better read outthentic documentaion first. 

Regards, Akeksei .

**Comment by Manwar** [comment]
Hi Alexey,

Thanks for the link to the documentation.
I will get back to you once I have absorbed the details.

Best Regards,
Mohammad S Anwar

**Comment by Manwar** [comment]
Hi Alexey,

I have a proposal for you. 

Can you please upgrade it to cover all the maps available on CPAN?

Once done then release the under than namespace Map::Tube::Tester to CPAN, so that other can benefit from your hard work and also submit PR if they want.

Many Thanks.

Best Regards,
Mohammad S Anwar

**Comment by melezhik** [comment]
Hi Mohammad!

> Once done then release it under than namespace Map::Tube::Tester to CPAN, so that other can benefit from your hard work and also submit PR if they want.

How do you see a use case for such Map::Tube::Tester cpan module? Who is going to use it?
Could you please provides some cases for it? Before go into realization I wish we could understand some practical meaning of it. There are some possible variants:

* trigger travis jobs which hit map-tube-testers tests against a brand new Map::Tube::* module code
* let other Map::Tube::* developers to `git clone  map-tube-testers` and then somehow run tests against their current Map::Tube::* code
* making a real Map::Tube::Tester module which encapsulate all the testing logic, but how one could use it? Another drawback behind having tests packed into cpan modules is PAUSE index gets updated quite slow, in case we need to change something in test logic quickly we'd better look towards keeping tests on github and running them over `git clone && carton exec` ( previous two points ). Tough packaging tests into cpan module has profits as well. I just can't see _how_ we gonna use it, still, may be you have better ideas ...

* something else ?

**Comment by Manwar** [comment]
Hi Michal,

Apology for late response.

I am going to be the first user of Map::Tube::Testers and hopefull others, who have map code would also join the group as well.

Ideally what I am looking for is, the centralised point of testing all maps. For example, everytime I make changes to core module i.e. Map::Tube, it would be great if I can test the change against all the maps dependent on Map::Tube before making the changes public.

In the past, there have been instances where I made changes and it broke other users map but not mine. 

To kick start the project, I would suggest collect all existing test cases from every maps available and put in format that your tool expects. Then the user can run all or just a particular map depending on his requirement.

This would also make the map lighter as well since all map testing code is in now centralised.

Have I made myself clear?

Please give me a shout if you have any queries. Please bear with me if I don't reply promptly.

Many Thanks.

Best Regards,
Mohammad S Anwar

**Comment by melezhik** [comment]
Hi Mohammad. Let me think about. As far as I know there are a plenty of existed maps, not sure if I am able to cover all the stuff shortly ...

**Comment by melezhik** [comment]
further discussion moved there - https://github.com/melezhik/map-tube-testers , has emailed you.
